src/daemon/http/request.rs

```diff
@@ -0,0 +1,527 @@
+    pub fn uri(&self) -> &Uri {
+        self.request.uri()
+    }
+    */
+
+    /// Returns the current request path.
+    pub fn path(&self) -> Result<RequestPath, InvalidPath> {
+        RequestPath::from_request(self)
+    }
+
+    /// Returns a reference to the Hyper request.
+    pub fn hyper(&self) -> &HyperRequest {
+        &self.request
+    }
+
+    /*
+    /// Returns the headers of the request.
+    pub fn headers(&self) -> &HeaderMap {
+        self.request.headers()
+    }
+    */
+
+    /// Returns the user agent header if present.
+    pub fn user_agent(&self) -> Option<String> {
+        match self.request.headers().get(&USER_AGENT) {
+            None => None,
+            Some(value) => value.to_str().ok().map(|s| {
+                // Note: HeaderValue.to_str() only returns ok in case the
+                // value is plain       ascii so it's safe to
+                // treat bytes as characters here.
+                if s.len() > HTTP_USER_AGENT_TRUNCATE {
+                    s[..HTTP_USER_AGENT_TRUNCATE].to_string()
+                } else {
+                    s.to_string()
+                }
+            }),
+        }
+    }
+
+    /// Checks for permissions.
+    ///
+    /// Returns an appropriate error response if the permissions are not met.
+    pub fn check_permission(
+        &self, permission: Permission, resource: Option<&MyHandle>
+    ) -> Result<(), HttpResponse> {
+        self.auth.check_permission(permission, resource).map_err(|err| {
+            HttpResponse::response_from_error(Error::from(err))
+        })
+    }
+
+    /// Checks the permissions and progresses to the next processing stage.
+    ///
+    /// If the authentication information for the request has the given
+    /// permissions, returns an [`AuthedRequest`] and an [`Actor`] which
+    /// allow further processing.
+    ///
+    /// Otherwise returns an appropriate error response.
+    pub fn proceed_permitted(
+        self, 
+        permission: Permission,
+        resource: Option<&MyHandle>,
+    ) -> Result<(AuthedRequest<'a>, AuthInfo), HttpResponse> {
+        self.check_permission(permission, resource)?;
+        Ok((
+            AuthedRequest {
+                request: self.request,
+                server: self.server,
+                limits: self.limits,
+            },
+            self.auth
+        ))
+    }
+
+    /// Permits the request to the next processing stage.
+    ///
+    /// Returns [`AuthedRequest`] and [`Actor`] without requiring any
+    /// permissions whatsoever.
+    pub fn proceed_unchecked(
+        self
+    ) -> (AuthedRequest<'a>, AuthInfo) {
+        (
+            AuthedRequest {
+                request: self.request,
+                server: self.server,
+                limits: self.limits,
+            },
+            self.auth
+        )
+    }
+
+    /// Splits the request into the server and raw Hyper request.
+    pub fn proceed_raw(self) -> (&'a HttpServer, HyperRequest) {
```

Comment:
This doesn't feel like it should be called proceed because it doesn't do anything, this is just an into fn, maybe call it into_parts() or something?

Reply:
After looking at `proceed_unchecked()` and thinking that could be called `into_authenticated_request()` maybe call this `into_raw_request()` ?

Reply:
I found `proceed_permitted` already too long.

I picked _proceed_ here as meaning “continue with what you were doing after a brief interruption,” in this case for the permission check, and then decided to keep it persistently. I don’t like using `into` since some of those functions actually do something more than just conversion.